File: apps/api/src/routes/update.ts
import { Hono } from 'hono'
import { fail, ok } from '../lib/responses.js'
import { updateService } from '../services/update.service.js'

const app = new Hono()

app.get('/update', async (c) => {
  return ok(c, await updateService.getOverview())
})

app.get('/update/runs/:id', async (c) => {
  const run = await updateService.getRun(c.req.param('id'))
  if (!run) {
    return fail(c, 'Update run not found', 404)
  }

  return ok(c, run)
})

app.post('/update/check', async (c) => {
  try {
    return ok(c, await updateService.forceCheck())
  } catch (error) {
    return fail(c, error instanceof Error ? error.message : 'Failed to refresh releases', 500)
  }
})

app.post('/update/runs', async (c) => {
  try {
    return ok(c, await updateService.createRunForLatestRelease(), 201)
  } catch (error) {
    return fail(c, error instanceof Error ? error.message : 'Failed to queue update', 409)
  }
})

app.post('/update/runs/:id/retry', async (c) => {
  try {
    return ok(c, await updateService.retryRun(c.req.param('id')))
  } catch (error) {
    return fail(c, error instanceof Error ? error.message : 'Failed to retry update', 409)
  }
})

app.post('/update/accept', async (c) => {
  try {
    return ok(c, await updateService.acceptLatestRelease())
  } catch (error) {
    return fail(c, error instanceof Error ? error.message : 'Failed to start update', 409)
  }
})

app.post('/update/decline', async (c) => {
  try {
    await updateService.declineLatestRelease()
    return ok(c, await updateService.getOverview(true))
  } catch (error) {
    return fail(c, error instanceof Error ? error.message : 'Failed to dismiss release', 500)
  }
})

export default app

File: backend/src/clawbuddy/routers/update.py
"""Update router — self-update management via Docker Swarm.

Replaces: apps/api/src/routes/update.ts
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter

from clawbuddy.lib.responses import fail, ok

router = APIRouter(tags=["update"])


@router.get("/update")
async def get_update_overview() -> Any:
    """Get current update status overview."""
    from clawbuddy.services.update import update_service

    return ok(await update_service.get_overview())


@router.post("/update/check")
async def check_for_updates() -> Any:
    """Force-check for new releases."""
    try:
        from clawbuddy.services.update import update_service

        return ok(await update_service.force_check())
    except Exception as exc:
        return fail(str(exc) or "Failed to refresh releases", status_code=500)


@router.post("/update/accept")
async def accept_update() -> Any:
    """Accept and start the latest release update."""
    try:
        from clawbuddy.services.update import update_service

        await update_service.accept_latest_release()
        return ok(await update_service.get_overview(force_release_refresh=True))
    except Exception as exc:
        return fail(str(exc) or "Failed to start update", status_code=409)


@router.post("/update/decline")
async def decline_update() -> Any:
    """Dismiss/decline the latest release."""
    try:
        from clawbuddy.services.update import update_service

        await update_service.decline_latest_release()
        return ok(await update_service.get_overview(force_release_refresh=True))
    except Exception as exc:
        return fail(str(exc) or "Failed to dismiss release", status_code=500)
